source/common/router/retry_state_impl.cc

```diff
@@ -374,41 +414,82 @@ bool RetryStateImpl::wouldRetryFromHeaders(const Http::ResponseHeaderMap& respon
            (retry_on_ & RetryPolicy::RETRY_ON_GRPC_UNAVAILABLE)) ||
           (status.value() == Grpc::Status::Internal &&
            (retry_on_ & RetryPolicy::RETRY_ON_GRPC_INTERNAL))) {
-        return true;
+        return RetryDecision::RetryWithBackoff;
       }
     }
   }
 
-  return false;
+  return RetryDecision::NoRetry;
 }
 
-bool RetryStateImpl::wouldRetryFromReset(const Http::StreamResetReason reset_reason) {
+RetryState::RetryDecision
+RetryStateImpl::wouldRetryFromReset(const Http::StreamResetReason reset_reason,
+                                    AlternateProtocolsUsed alternate_protocols_used,
+                                    bool& disable_alternate_protocols) {
+  ASSERT(!disable_alternate_protocols);
   // First check "never retry" conditions so we can short circuit (we never
   // retry if the reset reason is overflow).
   if (reset_reason == Http::StreamResetReason::Overflow) {
-    return false;
+    return RetryDecision::NoRetry;
+  }
+
+  const bool consider_disable_alt_svc = Runtime::runtimeFeatureEnabled(
+      "envoy.reloadable_features.conn_pool_new_stream_with_early_data_and_alt_svc");
+  if (consider_disable_alt_svc) {
+    if (reset_reason == Http::StreamResetReason::ConnectionFailure) {
+      if (alternate_protocols_used != AlternateProtocolsUsed::Unknown) {
+        // Already got request encoder, so this must be a 0-RTT handshake failure. Retry
+        // immediately.
+        // TODO(danzh) consider making the retry configurable.
+        ASSERT(alternate_protocols_used == AlternateProtocolsUsed::Yes,
+               "0-RTT was attempted on non-Quic connection and failed.");
+        return RetryDecision::RetryImmediately;
+      }
+      if ((retry_on_ & RetryPolicy::RETRY_ON_CONNECT_FAILURE)) {
+        // This is a pool failure.
+        return RetryDecision::RetryWithBackoff;
+      }
+    } else if (alternate_protocols_used == AlternateProtocolsUsed::Yes &&
+               (retry_on_ & RetryPolicy::RETRY_ON_ALT_PROTOCOLS_POST_CONNECT_FAILURE)) {
+      // Retry any post-handshake failure immediately with no alt_svc if the request was sent over
+      // Http/3.
```

Comment:
So this strikes me as a bit sketchy. 
If there's a way for an upstream to communicate "I don't support HTTP/3" blacklisting HTTP/3 make sense.  But for explicit-http3 config, saying "we got an early error so we're going to ignore the standard retry config, retry and fast-fail" seems sub-optimal behavior, because it'll burn a retry just determining that the grid doesn't support alt-protocols failover, if that makes sense?

Reply:
If a user configures explicit-http3 pool, IMO they shouldn't config RETRY_ON_ALT_PROTOCOLS_POST_CONNECT_FAILURE, knowing their pool doesn't have fallback. So no retry for H3 post handshake failure. WDYT?

Reply:
I made it not depending on the config but retry only with auto pool. PTAL